### ISSUE TITLE
[2.19] Update documentation on Dart library directive to include non-name libraries

### ISSUE THREAD

**Issue by kevmoo** [dart-lang/site-www#4314]
See https://github.com/dart-lang/language/issues/1073

Ask @srawlins for clarification

**Comment by srawlins** [comment]
To give a little summary here, because there's really nothing to this, haha:

We already have a "library directive" which can be written above imports, like `library my_library;`. You can use this identifier in part files (in an older style), or hang annotations off it (like `@deprecated`), or hang doc comments off it (with `/// comment`). The new feature is just that you don't need to spell out a name; you can write `library;`, hang annotations and doc comments off of that, and no need to write the name.

Previously the name was used in part files, and in mirrors, both of which are older style / soft-deprecated.

So an exciting use for this feature is really just some code like 

```dart
/// A really great test library.
@TestOn('browser')
library;

...
```

**Comment by mit-mit** [comment]
I think the documentation impact might be simply checking if we have any library directives in our docs that have annotations or doc comments, and if so, update those to use use the new `library;` syntax?

**Comment by kevmoo** [comment]
@mit-mit – likely. Just good to double-check.

**Comment by MaryaBelanger** [comment]
I have a question: Is there a reason to explicitly name libraries _at all_ anymore? 
Should we just completely discourage naming libraries and only ever show examples of `library;` _unnamed_?

[One note in the docs](https://dart.dev/guides/libraries/create-library-packages#what-makes-a-library-package:~:text=Note%3A%20When%20the%20library%20directive%20isn%E2%80%99t%20specified%2C%20a%20unique%20tag%20is%20generated%20for%20each%20library%20based%20on%20its%20path%20and%20filename.%20Therefore%2C%20we%20suggest%20that%20you%20omit%20the%20library%20directive%20from%20your%20code%20unless%20you%20plan%20to%20generate%20library%2Dlevel%20documentation.) says:

> "When the library directive isn’t specified, **_a unique tag is generated for each library_** based on its path and filename." 

I'm assuming that is now the case for unnamed `library;` declarations as well, so it seems like there's no reason to name them. 

It also seems like the only use for named libraries might be `part of`, which is old and not used anyway @srawlins?. We could get rid of this [Effective Dart section](https://dart.dev/guides/language/effective-dart/usage#do-use-strings-in-part-of-directives), since we literally say we [recommend avoiding it](https://dart.dev/guides/libraries/create-library-packages#what-makes-a-library-package:~:text=Note%3A%20You%20may%20have%20heard%20of%20the%20part%20directive%2C%20which%20allows%20you%20to%20split%20a%20library%20into%20multiple%20Dart%20files.%20We%20recommend%20that%20you%20avoid%20using%20part%20and%20create%20mini%20libraries%20instead.) in another note. 

In another thread [Lasse says](https://github.com/dart-lang/language/issues/1073#issue-652398032): "Giving a library a name is only really useful for avoiding duplicate imports with different paths, so most libraries do not use the declaration." But if that's really the only reason for naming libraries, wouldn't a unique URI would bypass that anyway?

I'm asking because information about the library directive is spread out all over the place and it's pretty inefficient to write "you can name them but don't need to unless you want to use `part of`, which you shouldn't really be doing anyway" everywhere it's mentioned.

**Comment by kevmoo** [comment]
@MaryaBelanger – not really.

If you're using `dart:mirrors` – maybe. But we discourage that.
You don't need `library` for parts, either. You SHOULD use the URI of the library instead of the library name.

The ONLY (good) reason for library is `///` doc comments and/or `@annotations`.

**Comment by kevmoo** [comment]
See https://dart-lang.github.io/linter/lints/unnecessary_library_directive.html !!

**Comment by MaryaBelanger** [comment]
Ok might be a little drastic but I'm going to add an Effective Dart: Usage section that says something like "DO only use the `library` directive for doc comments and annotations". It'd be really nice to point to one place for that info (that's not the linter docs, which is the most complete source atm since it's already updated for 2.19)

**Comment by kevmoo** [comment]
Please do! Then we should update the lint to reference that this is a style guide "thing" – @srawlins @pq

**Comment by kevmoo** [comment]
Yeah there's a linter rule. That's great. But we should update the style guide, too.

**Comment by mit-mit** [comment]
cc @munificent

**Comment by munificent** [comment]
Honestly, it's probably not even worth adding to "Effective Dart". There are an infinite number of bad things we could tell users not to do but if the bad thing takes effort to do and doesn't have any even perceived upside, we don't usually tell them not to because they just... won't. :)

We can add a guideline against it if others think it's useful, but I'm always a little worried about "Effective Dart" getting so many rules that people just tune it out entirely.

**Comment by MaryaBelanger** [comment]
> There's already a linter rule (that words it much better!): https://dart.dev/tools/linter-rules#unnecessary_library_directive

So this linter rule cites the existing Effective Dart: Documentation section "DO use library directives if you want to document a library and/or annotate a library."

But I noticed there's not a linter rule that's just like `unnecessary_library_name` or something. Should that be a thing too @srawlins ?
I'm asking because of this line (that was already in the docs before I started editing):
> A unique tag is generated for each library based on its path and filename. Naming libraries overrides their intrinsic URI, **which can actually make it harder for tools to physically find the main library** file you’re referring to. 

I feel like this is a different concern than what the `unnecessary_library_directuve` lint addresses.

There's also this lint, `library_names` which at the very least is irrelevant now because naming libraries is useless. Should it be removed, or changed to focus on useless names and not the syntax of names?

**Comment by MaryaBelanger** [comment]
@munificent Good point. 

1. I will remove the ["DON'T use library directives unless..."](https://github.com/dart-lang/site-www/pull/4390/files#diff-7ff65f715c2ef8a4221fd003bf2ca2d543ff0dcd7e881a869adf36c761b9491cR24) section I added. (_especially since that content is already more or less completely covered in the existing ["CONSIDER writing library level doc comments"](https://dart.dev/guides/language/effective-dart/documentation#consider-writing-a-library-level-doc-comment) section_).

2. I want to keep the other section I added, ["DON'T explicitly name libraries"](https://github.com/dart-lang/site-www/pull/4390/files#diff-7ff65f715c2ef8a4221fd003bf2ca2d543ff0dcd7e881a869adf36c761b9491cR51), **_just in case_** there is a linter rule to be added (see [my previous comment](https://github.com/dart-lang/site-www/issues/4314#issuecomment-1335841146)).  But, I'll move it to Effective Dart: Style, instead of Usage, since that's what linter rules point to. 

    **_If a linter rule for "don't name libraries" doesn't get created, I'll remove it completely._** 

3. Idk why I didn't think of this before, but I think the minutiae of library directive names could actually be a short subsection under [Libraries and visibility](https://dart.dev/guides/language/language-tour#libraries-and-visibility), so I'm going to add that. 

How does that sound?

**Comment by pq** [comment]
LGTM.

> If a linter rule for "don't name libraries" doesn't get created, I'll remove it completely.

Tracking issue: https://github.com/dart-lang/linter/issues/3882.

**Comment by MaryaBelanger** [comment]
Fixing PRs merged into [v2.19](https://github.com/dart-lang/site-www/tree/v2.19)
